go/plugins/googlegenai/gemini.go

```diff
@@ -69,6 +72,265 @@ type EmbedOptions struct {
+	ResponseLogprobs bool     `json:"responseLogprobs,omitempty"`
+	Logprobs         *int32   `json:"logprobs,omitempty"`
+	PresencePenalty  *float32 `json:"presencePenalty,omitempty"`
+	FrequencyPenalty *float32 `json:"frequencyPenalty,omitempty"`
+	Seed             *int32   `json:"seed,omitempty"`
+
+	// Safety settings
+	SafetySettings []*SafetySetting `json:"safetySettings,omitempty"`
+
+	// Model version
+	Version string `json:"version,omitempty"`
+}
+
+// All the supporting types needed by GeminiConfig:
+
+type Tool struct {
+	FunctionDeclarations []*FunctionDeclaration `json:"functionDeclarations,omitempty"`
+}
+
+type FunctionDeclaration struct {
+	Name        string         `json:"name,omitempty"`
+	Description string         `json:"description,omitempty"`
+	Parameters  map[string]any `json:"parameters,omitempty"`
+}
+
+type ToolConfig struct {
+	FunctionCallingConfig *FunctionCallingConfig `json:"functionCallingConfig,omitempty"`
+}
+
+type FunctionCallingConfig struct {
+	Mode                 string   `json:"mode,omitempty"`
+	AllowedFunctionNames []string `json:"allowedFunctionNames,omitempty"`
+}
+
+type SpeechConfig struct {
+	Voice string `json:"voice,omitempty"`
+}
+
+type ResponseSchema struct {
+	Type       string                    `json:"type,omitempty"`
+	Properties map[string]SchemaProperty `json:"properties,omitempty"`
+	Required   []string                  `json:"required,omitempty"`
+}
+
+type SchemaProperty struct {
+	Type        string                    `json:"type,omitempty"`
+	Description string                    `json:"description,omitempty"`
+	Items       *SchemaProperty           `json:"items,omitempty"`
+	Properties  map[string]SchemaProperty `json:"properties,omitempty"`
+}
+
+type Content struct {
+	Parts []*ContentPart `json:"parts,omitempty"`
+	Role  string         `json:"role,omitempty"`
+}
+
+type ContentPart struct {
+	Text       string      `json:"text,omitempty"`
+	InlineData *InlineData `json:"inlineData,omitempty"`
+}
+
+type InlineData struct {
+	MimeType string `json:"mimeType,omitempty"`
+	Data     string `json:"data,omitempty"`
+}
+
+type RoutingConfig struct {
+	ModelRoutes []*ModelRoute `json:"modelRoutes,omitempty"`
+}
+
+type ModelRoute struct {
+	ModelName string         `json:"modelName,omitempty"`
+	Criteria  *ModelCriteria `json:"criteria,omitempty"`
+}
+
+type ModelCriteria struct {
+	TextLength int32 `json:"textLength,omitempty"`
+	MediaCount int32 `json:"mediaCount,omitempty"`
+}
+
+type ThinkingConfig struct {
+	Enabled bool `json:"enabled,omitempty"`
+}
+
+type HTTPOptions struct {
+	Timeout int `json:"timeout,omitempty"`
+}
+
+// extractConfigFromInput converts any supported config type to GoogleAIConfig
+func extractConfigFromInput(input *ai.ModelRequest) (GeminiConfig, error) {
+	var result GeminiConfig
+	switch config := input.Config.(type) {
+	case GeminiConfig:
+		return config, nil
+	case *GeminiConfig:
+		return *config, nil
+	case ai.GenerationCommonConfig:
```

Comment:
Instead of duplicating this logic, can we either have a helper function or just detect that it is/isn't a pointer and take the pointer of it? Doesn't seem necessary to be this many lines.